Validate resource usage flags when used

yanchith opened the pull request:
This PR adds usage flag validations to the functions discussed in https://github.com/gfx-rs/wgpu/issues/228#issuecomment-528949445 and https://github.com/gfx-rs/wgpu/issues/228#issuecomment-528953096

I added `usage` fields to `Buffer` and `Texture` for convenience.

The validations themselves are run before much else happens (seemed clearer that way). This means certain `VecMap`s and `FastHashMap`s might be queried more than necessary, though. 

I also left one TODO in, which I'd like to resolve: With what usage flags should the swapchain textures be created? Currently it initializes the texture resource trackers with `UNINITIALIZED`, so I just copied the same flag for the texture. This just happens to work (that is not trigger the validations I added), because its value, `UNINITIALIZED`, pretends to be every usage flag by having all bits set. Is that by design?

Closes #228 ~(except for `BufferUsage::INDIRECT`, for which I can create another issue)~

kvark commented on wgpu-native/src/command/mod.rs:
```diff
@@ -249,20 +249,42 @@ pub fn command_encoder_begin_render_pass<B: GfxBackend>(
             );
 
-            if let Some(resolve) = unsafe { at.resolve_target.as_ref() } {
+            let color_attachment_texture = &texture_guard[color_attachment_view.texture_id.value];
```
Since we access the texture a little bit later anyway, let's move the check there in order to avoid fetching the texture data storage more than necessary. Doing the checks as early as possible is a worthy goal, but only when it doesn't harm our performance.

kvark commented on wgpu-native/src/command/mod.rs:
```diff
@@ -249,20 +249,42 @@ pub fn command_encoder_begin_render_pass<B: GfxBackend>(
+            let color_attachment_texture = &texture_guard[color_attachment_view.texture_id.value];
+            assert!(
+                color_attachment_texture.usage.contains(TextureUsage::OUTPUT_ATTACHMENT),
+                "Color attachment texture must be created with TextureUsage::OUTPUT_ATTACHMENT",
```
Given the proper variable names, the assertion is pretty self-describing through the code. The comment seems to be just repeating the same thing, so it doesn't look necessary.

yanchith replied:
👍  I didn't realize `assert!` actually gives a pretty decent error message (unlike `assert_eq!`)

kvark commented on wgpu-native/src/command/render.rs:
```diff
@@ -363,6 +369,12 @@ pub fn render_pass_set_vertex_buffers<B: GfxBackend>(
         .iter_mut()
         .zip(buffers.iter().zip(offsets))
     {
+        let buffer_usage = buffer_guard[id].usage;
```
we get the buffer a few lines later, let's move the check there in order to avoid `buffer_guard` access here

kvark commented on wgpu-native/src/command/transfer.rs:
```diff
@@ -81,6 +81,17 @@ pub fn command_encoder_copy_buffer_to_buffer<B: GfxBackend>(
     // borrow the buffer tracker mutably...
     let mut barriers = Vec::new();
 
+    let src_buffer_usage = buffer_guard[source].usage;
```
similarly, let's use the `src_buffer.usage` and `dst_buffer.usage` when it's available, no need to rush for this check

kvark commented on wgpu-native/src/device.rs:
```diff
@@ -1182,19 +1184,33 @@ pub fn device_create_bind_group<B: GfxBackend>(
+                    let buffer_usage = buffer_guard[bb.buffer].usage;
                     let (alignment, usage) = match decl.ty {
                         binding_model::BindingType::UniformBuffer => {
+                            assert!(
```
we don't need to do the checks right in these arms: we get `usage` as a result of the match, and `buffer` a few lines afterwards, so we can just do a single check at that point

kvark commented on wgpu-native/src/device.rs:
```diff
@@ -2018,6 +2050,8 @@ pub fn device_create_swap_chain<B: GfxBackend>(
                 ref_count: device.life_guard.ref_count.clone(),
             },
+            // TODO: what should be here? UNINITIALIZED or OUTPUT_ATTACHMENT? Or both?
+            usage: resource::TextureUsage::UNINITIALIZED,
```
let's have `OUTPUT_ATTACHMENT` only

yanchith replied:
Also in the tracker a few lines below?

kvark replied:
right, the tracker should still consider it UNITIALIZED

yanchith commented:
Thanks! I originally had the checks a bit down, without accessing the data multiple times. I'll move them back 🙂

yanchith commented:
Done!

One last thing I am not very sure about. In `command_encoder_begin_render_pass()` we use `trackers.textures.query()` and only fetch (and validate) the texture in the `None` case. AIUI the `None` case happens the first time the resource is referenced in the command buffer. Is that right?

kvark commented on wgpu-native/src/device.rs:
```diff
@@ -1192,9 +1194,7 @@ pub fn device_create_bind_group<B: GfxBackend>(
-                        binding_model::BindingType::Sampler
-                        | binding_model::BindingType::SampledTexture
-                        | binding_model::BindingType::StorageTexture => {
+                        _ => {
```
please don't change this - it's unrelated, and we explicitly want to enumerate the variants

kvark commented:
> One last thing I am not very sure about. In command_encoder_begin_render_pass() we use trackers.textures.query() and only fetch (and validate) the texture in the None case. AIUI the None case happens the first time the resource is referenced in the command buffer. Is that right?

Great question! I think we should split the `use_replace` that is in the `None` arm into getting the texture and then changing the sate. This way, we'd move the former up outside the match and have the usage assert near it, while the latter will be done inside the match arm. Please let me know if you find this explanation confusing.

yanchith commented:
Okay! I split it out getting the texture and the assert and moved it before the match statement.

Btw what are the trackers for? Inserting synchronization primitives when something changed with the resource that is not synchronized by default? Anywhere I could read up on it a bit?

yanchith commented:
FYI the CI fail is nightly only

kvark commented:
Great, thank you!
Given some of the back-and-forth in the commits, could you squash them please?

yanchith commented:
Will do!
